lint/rule.go

```diff
@@ -14,10 +14,172 @@
+			}
+		}
+		for _, enum := range message.GetNestedEnumTypes() {
+			if rule.LintEnum != nil {
+				problems = append(problems, rule.LintEnum(enum)...)
+			}
+		}
+	}
+
+	// Iterate over each service and process any rules within that service,
+	// as well as rules for each method.
+	for _, service := range f.GetServices() {
+		if rule.LintService != nil {
+			problems = append(problems, rule.LintService(service)...)
+		}
+		for _, method := range service.GetMethods() {
+			if rule.LintMethod != nil {
+				problems = append(problems, rule.LintMethod(method)...)
+			}
+		}
+	}
+
+	// Process rules for top-level enums.
+	for _, enum := range f.GetEnumTypes() {
+		if rule.LintEnum != nil {
+			problems = append(problems, rule.LintEnum(enum)...)
+		}
+	}
+
+	// Finally, process rules for the file itself.
+	if rule.LintFile != nil {
+		problems = append(problems, rule.LintFile(f)...)
+	}
+
+	// Done aggregating problems for this file; return the list.
+	return problems
+}
+
+// IsEnabled returns true if the rule is enabled (not disabled by the comments
+// for the given descriptor or its file), false otherwise.
+func (rule *Rule) IsEnabled(d desc.Descriptor) bool {
+	directive := fmt.Sprintf("api-linter: %s=disabled", rule.Name)
+
+	// If the comments above the descriptor disable the rule,
+	// return false.
+	if sourceInfo := d.GetSourceInfo(); sourceInfo != nil {
+		if strings.Contains(sourceInfo.GetLeadingComments(), directive) {
+			return false
+		}
+	}
+
+	// The rule may also be disabled at the file level.
+	// If it is, return false.
+	if strings.Contains(fileHeader(d.GetFile()), directive) {
+		return false
+	}
+
+	// The rule is enabled.
+	return true
+}
+
+// getAllMessages returns a slice with every message (not just top-level
+// messages) in the file.
+func getAllMessages(f *desc.FileDescriptor) (messages []*desc.MessageDescriptor) {
+	messages = append(messages, f.GetMessageTypes()...)
+	for _, message := range f.GetMessageTypes() {
+		messages = append(messages, getAllNestedMessages(message)...)
+	}
+	return messages
+}
+
+// getAllNestedMessages returns a slice with the given message descriptor as well
+// as all nested message descriptors, traversing to arbitrary depth.
+func getAllNestedMessages(m *desc.MessageDescriptor) (messages []*desc.MessageDescriptor) {
+	for _, nested := range m.GetNestedMessageTypes() {
+		messages = append(messages, nested)
+		messages = append(messages, getAllNestedMessages(nested)...)
+	}
+	return messages
+}
+
+// fileHeader attempts to get the comment at the top of the file, but it
+// is on a best effort basis because protobuf is inconsistent.
+//
+// Taken from https://github.com/jhump/protoreflect/issues/215
+func fileHeader(fd *desc.FileDescriptor) string {
+	var firstLoc *descriptor.SourceCodeInfo_Location
+	var firstSpan int64
+	for _, curr := range fd.AsFileDescriptorProto().GetSourceCodeInfo().GetLocation() {
+		if curr.LeadingComments == nil && len(curr.LeadingDetachedComments) == 0 {
```

Comment:
I think it will make some mistakes in the following case:
```protobuf
syntax "proto3";

// This is not the file header, but it will be picked up by this function 
// and the following rule will be disabled for all messages in the file.
// api-linter: check-message-name = disabled
message M {}
```

So please add some tests for this function.

Reply:
See jhump's comment in the original issue.

Reply:
Updated, and test added.